File: BreadthFirstSearch/BreadthFirstSearch/Sources/Graph.swift
public class Edge {
  public var neighbor: Node

  init(_ neighbor: Node) {
    self.neighbor = neighbor
  }
}

public class Node {
  public var visited: Bool
  public var neighbors: [Node]
  public var label: String

  init(label: String) {
    self.label = label
    visited = false
    neighbors = []
  }

  public func addNeighbor(_ node: Node) -> Void {
    self.neighbors.append(node)
  }
}

// public class Graph {
//   public var nodes: [Node] = []

//   public func addNode(_ label: String) -> Node {
//     let node = Node(label: label)
//     nodes.append(node)
//     return node
//   }

//   public func addEdge(_ source: Node, neighbor: Node) {
//     let edge = Edge(neighbor)
//     source.edges.append(edge)
//   }
// }

File: DepthFirstSearch/depth-first-search/src/depth_first_search/graph.clj
(ns depth-first-search.graph)

(defn mark-visited [graph source]
  (assoc-in graph [source :visited] true))

(defn add-node [nodes label]
  (assoc nodes label {:neighbors []}))

(defn add-edge [graph source neighbor]
  (let [neighbors-of (comp :neighbors graph)]
    (assoc graph source {:neighbors (conj (neighbors-of source) neighbor)})))
